RFCs/0009-Fn::Map.md

```diff
@@ -0,0 +1,898 @@
+# New Intrinsic Function for looping: `Fn::Map`
+
+* **Original Author(s):**: @mingujo
+* **Tracking Issue**: [Tracking Issue: Fn::Map](https://github.com/aws-cloudformation/cfn-language-discussion/issues/9)
+
+# Summary
+We are proposing to introduce a new intrinsic function called `Fn::Map`. It will allow users to iterate over a given list to replicate a particular template fragment thereby avoiding repetitively writing a similar declaration for each object (e.g. Resource, Output, or Resource Properties).
+
+`Fn::Map` traverses a single given List of Strings from first to last, and over each iteration, it references each item under a given template snippet to be replicated.
+
+We additionally introduce `Fn::Merge` which can be used to merge the result of Fn::Map into its parent object.
+# Motivation
+
+In a CloudFormation template, a single Resource configures one infrastructure object. Therefore, a template can become verbose when you manually declare multiple similar resources. Some examples would include a pool of EC2 instances with the same configurations but different instance types, or S3 Bucket Notification configurations with different SNS Topics.
+
+# Declaration
+## Fn::Map
+### Json
+```json
+{
+  "Fn::Map": {
+    "Index": "Index",
+    "Value": "Value",
+    "Collection": "CollectionToIterate",
+    "Fragment": { },
+    "Key": "Key"
+  }
+}
+```
+
+### Yaml
+```yaml
+Fn::Map: # or !Map
+  Index: "Index"
+  Value: "Value"
+  Collection: "CollectionToIterate"
+  Fragment: { }
+  Key: "Key"
+```
+
+### Parameters
+
+Fn::Map supports 5 parameters.
+
+* `Index` (optional)
+    * The name of an iterator variable which stores a positional index of each value in a given List.
+    * The index number sequence starts from 0.
+    * String type
+    * the default value is "Index"
+* `Value` (optional)
+    * The name of an iterator variable which stores each value in a List.
+    * String type
+    * the default value is "Value"
+* `Collection`
+    * The collection to iterate on. It must be a List. You can either in-place the collection itself or reference List type parameter values from the Parameters section.
+* `Fragment`
+    * The fragment to replicate for each item in iteration
+* `Key` (conditional)
+    * The json or yaml key for replicated objects. For Resources and Outputs this will be the logical id.
+    * **This field is only required when Fn::Map should generate an object of key value pairs, for example in the Resources or Outputs section of a template. It should not be specified when using Fn::Map to generate an array, for example a list type attribute for a Resource Property.** 
+    * The value **should be interpolated with an index** using `Fn::Sub` intrinsic function.
+
+
+## Fn::Merge
+Fn::Merge takes an array of objects as input and merges them into the parent object
+```json
+{
+  "Fn::Merge": [
+    {
+      "key1": { },
+      "key2": { }
```

Comment:
Let's capitalize these as well to remain consistent, even if these aren't real values.

Reply:
Sure, we might as well